### DISCUSSION TITLE
Custom asserts

### DISCUSSION THREAD

**Discussion by tpluscode** [Orange-OpenSource/hurl#1032] [Ideas]
Is it possible to create more asserts to handle different response bodies in specific ways?

**Comment by jcamiel** [comment]
Currently no, but the last entry of an Hurl file outputs the HTTP response so you can pipe this response to a script of your choice and add your custom logic.
Can you share uses cases for the asserts you would like to use so we can have these as inputs ?

**Reply by tpluscode** [reply]
Piping the last request is not enough if I would like to use hurl in `--test` mode. Possibly from multiple sources like `hurl --test test/**/*.hurl`

I would propose a way to allow custom scripts to be injected in the `Body` section. In the past I experimented with IntelliJ's HTTP Client which has the [script handler](https://www.jetbrains.com/help/idea/http-client-in-product-code-editor.html#using-response-handler-scripts) functionality. Unfortunately, it run a very limited JavaScript engine

So, for starters, I would propose "full control" mode, which would allow something like

```
GET https://example.com

HTTP/* 200

> node ./com-script.js

GET https://example.org

HTTP/* 200

> node ./org-script.js
```

This would be equivalent to piping each request's body to a script, instead of just the last. Like

```
curl https://example.com | node ./com-script.js
curl https://example.org | node ./org-script.js
```

This was I am free to choose any language I want as long as I can consume stdin. If exit code is not 0, the stdout would be used by hurl as the error for a given request.

WDYT?

**Reply by tpluscode** [reply]
Off the top of my head, it would probably be necessary to support variables. Our script would likely need the content-type, to know how to parse the body without guessing

```
GET https://example.com/resource

HTTP/* 200
[Captures]
format: header "content-type"

> node ./test-body.js --format {{format}}
```

**Reply by luolong** [reply]
Few reasons why adding custom assertions/code blocks would be desirable. 

1. Making a chain of requests where following requests get input from previous ones (not necessarily just the last request either). That implies capturing and using variables in requests.

2. Custom script (maybe based on embedded Denojs engine) could be used to execute complex test assertions and/or calculating variables for the next request.

**Reply by jcamiel** [reply]
Hi @luolong, regarding your first point, you can already captures variables from any requests and reuse it anywhere in your file (see [capturing](https://hurl.dev/docs/capturing-response.html)). For the second point, we really understand the need, (see [#1072](https://github.com/Orange-OpenSource/hurl/issues/1072#issuecomment-1346135666)), we've currently postponed it.

**Reply by tpluscode** [reply]
Hello @jcamiel. What is the current position on my original proposal?

**Reply by jcamiel** [reply]
Hi @tpluscode 

Really no strong feelings about adding this feature for the moment. On the syntax side (this is not an endorsement to add the feature), if we want to add a way of transform response output, I would use a new option within `[Options]` section to do this:


```hurl
GET https://foo.com
[Options]
output-cmd: python3 filter.py
HTTP 200
```

**Reply by SilenLoc** [reply]
@jcamiel FYI: When doing my talk this was a feature many wanted to have. Some kind of break out of hurl into python or something to do different things. 

Mostly around the topics: "if else", "send Kafka events", "resolving paged responses"
